File: web/src/ball.js
export default class Ball {
    constructor(radiusMult, color) {
        this.radiusMult = radiusMult;
        this.color = color;
    }

    // update the position/velocity of the ball
    update(step) {
        this.yv += this.game.gravity * step;

        // enforce max speed
        if (Math.abs(this.xv) > this.radius / 1.6) {
            this.xv = this.maxV * (this.xv / Math.abs(this.xv));
        }
        if (Math.abs(this.yv) > this.radius / 1.6) {
            this.yv = this.maxV * (this.yv / Math.abs(this.yv));
        }

        this.x += this.xv * step;
        this.y += this.yv * step;
    }

    // render the ball
    draw() {
        var ctx = this.game.ctx;
        
        ctx.fillStyle = this.color;
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radius, 0, 2 * Math.PI, false);
        ctx.fill();
        ctx.lineWidth = 1;
        ctx.strokeStyle = "#003300";
        ctx.stroke();
        
        this.drawXInd(ctx);
    }

    // render a horizontal position indicator if the ball is off-screen
    drawXInd(ctx) {
        let radius = 0;
        if (this.y - this.radius < -this.radius * 18) {
            radius = this.radius * (4 / 24);
        } else if (this.y - this.radius < -this.radius * 14) {
            radius = this.radius * (5 / 24);
        } else if (this.y - this.radius < -this.radius * 10) {
            radius = this.radius * (6 / 24);
        } else if (this.y - this.radius < -this.radius * 6) {
            radius = this.radius * (7 / 24);
        } else if (this.y - this.radius < -this.radius * 2) {
            radius = this.radius * (8 / 24);
        }

        if (radius > 0) {
            ctx.fillStyle = this.game.backgroundColor;
            ctx.beginPath();
            ctx.arc(this.x, this.radius / 2, radius, 0, 2 * Math.PI, false);
            ctx.fill();
            ctx.lineWidth = 2;
            ctx.strokeStyle = "#003300";
            ctx.stroke();
        }
    }
}

File: desktop/Ball.py
## Chris Leveille
## April 2019

import pygame, pygame.gfxdraw

class Ball():

	## Create new ball
	def __init__(self, radius, color):
		
		self.x = 0
		self.y = 0
		self.xv = 0
		self.yv = 0
		self.maxV = 15
		self.color = color
		self.radius = radius

	## Update the ball's position
	def updatePosition(self, gravity):
		
		self.yv += gravity

		# Enforce the ball's maximum speed
		if abs(self.xv) > self.maxV:
			self.xv = self.maxV * (self.xv / abs(self.xv))
		if abs(self.yv) > self.maxV:
			self.yv = self.maxV * (self.yv / abs(self.yv))

		self.x += self.xv
		self.y += self.yv

	## Draw the ball
	def draw(self, gameWin):
		
		pygame.gfxdraw.filled_circle(gameWin, int(self.x), int(self.y), int(self.radius), self.color)
		pygame.gfxdraw.aacircle(gameWin, int(self.x), int(self.y), int(self.radius), pygame.color.Color("black"))
		self.drawXInd(gameWin)

	# If the ball is off-screen, draw a marker indicating its horizontal location
	def drawXInd (self, gameWin):
		
		if self.y - self.radius < -450:
			radius =  4
		elif self.y - self.radius < -350:
			radius =  5
		elif self.y - self.radius < -250:
			radius =  6
		elif self.y - self.radius < -150:
			radius =  7
		elif self.y - self.radius < -50:
			radius =  8
		else:
			radius = 0

		if radius > 0:
			pygame.gfxdraw.aacircle(gameWin, int(self.x), 10, radius, pygame.color.Color("black"))
			pygame.gfxdraw.aacircle(gameWin, int(self.x), 10, radius - 1, pygame.color.Color("black"))